micrometer-core/src/main/java/io/micrometer/core/instrument/observation/ObservationOrTimerCompatibleInstrumentation.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2022 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.observation;
+
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationRegistry;
+import io.micrometer.observation.transport.RequestReplySenderContext;
+
+import java.util.function.Supplier;
+
+/**
+ * Abstracts instrumenting code with a {@link Timer} or an {@link Observation}. This can
+ * be useful for avoiding complexity and minimizing overhead when converting
+ * instrumentation that was previously instrumented with a {@link Timer} to being
+ * optionally instrumented with an {@link Observation}. This may be necessary where
+ * backwards compatibility is a concern such that you cannot require an
+ * {@link ObservationRegistry}. If there are no backwards compatibility concerns,
+ * generally direct instrumentation with {@link Observation} should be preferred. While
+ * this was designed for use internally in micrometer-core where we have this need, it may
+ * also be useful to other libraries with pre-existing {@link Timer}-based
+ * instrumentation. If an {@link ObservationRegistry} is provided that is not the no-op
+ * registry, an {@link Observation} will be used for instrumentation. Otherwise, a
+ * {@link Timer} will be used.
+ *
+ * @param <T> context type if Observation used for instrumentation
+ */
+public class ObservationOrTimerCompatibleInstrumentation<T extends Observation.Context> {
```

Comment:
Do you add `@since` on public classes?

Reply:
In general, yes. I was still kind of on the fence when writing this whether it should be meant as an internal to Micrometer class (but needs to be public to be used from other packages) or if we should support other projects using it. I think it may make sense for some other projects that aren't having a major version when upgrading to Micrometer 1.10 but want to support Observation instrumentation in a backwards-compatible way like we do with our instrumentation.
I'll add `@since` - thanks.